org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/preferences/ClientPreferences.java

```diff
@@ -402,6 +403,27 @@ public boolean isCompletionListItemDefaultsInsertTextFormatSupport() {
 			&& capabilities.getTextDocument().getCompletion().getCompletionList().getItemDefaults().contains("insertTextFormat");
 	}
 
+	public boolean isCompletionListItemDefaultsInsertTextModeSupport() {
+		return v3supported
+			&& capabilities.getTextDocument().getCompletion() != null
+			&& capabilities.getTextDocument().getCompletion().getCompletionList() != null
+			&& capabilities.getTextDocument().getCompletion().getCompletionList().getItemDefaults() != null
+			&& capabilities.getTextDocument().getCompletion().getCompletionList().getItemDefaults().contains("insertTextMode");
+	}
+
+	public boolean isCompletionItemInsertTextModeSupport() {
+		return v3supported
+			&& capabilities.getTextDocument().getCompletion() != null
+			&& capabilities.getTextDocument().getCompletion().getCompletionItem() != null
+			&& capabilities.getTextDocument().getCompletion().getCompletionItem().getInsertTextModeSupport() != null
+			&& (capabilities.getTextDocument().getCompletion().getCompletionItem().getInsertTextModeSupport().getValueSet().contains(InsertTextMode.AsIs)
+			|| capabilities.getTextDocument().getCompletion().getCompletionItem().getInsertTextModeSupport().getValueSet().contains(InsertTextMode.AdjustIndentation));
+	}
+
+	public InsertTextMode getCompletionItemInsertTextMode() {
+		return capabilities.getTextDocument().getCompletion().getInsertTextMode();
```

Comment:
A little confused here. The document says:

`The client's default when the completion item doesn't provide an insertTextMode property`.

Then what is the purpose of setting this client default behavior to each completion item? Looks like this is already the default behavior if it's not set.

My understanding for this part of spec:

- Client tells server the insertTextModes it supports, and the default mode the client uses.
- Server can set insertTextMode if some of the items need to use another mode which is not the default one.

For example:

Client supports `asIs` & `adjustIndentation`. Meanwhile, the default mode is `asIs`.

At server side, for some completion item like snippets, it hopes to adjust the indentation when the text edit is applied. At this time, `adjustIndentation` needs to be set.

Reply:
You're right. After speaking to @JessicaJHee , I think we agree that we only need code to support the example above.

client supports `asIs` & `adjustIntentation`, while the default is `asIs`. Since we prefer `adjustIntentation` everywhere, that's the only case where it makes sense to override the client and set `adjustIndentation` for the completion list `itemDefaults`. Otherwise, nothing to do.

Reply:
Thank you for your correction! I had misunderstood the original issue. After discussing with @rgrunber, I believe the idea is that we should only set `insertTextMode` in the case that you mention here: 

> Client supports `asIs` & `adjustIndentation`. Meanwhile, the default mode is `asIs`.
> 
> At server side, for some completion item like snippets, it hopes to adjust the indentation when the text edit is applied. At this time, `adjustIndentation` needs to be set.

In other words, only set `insertTextMode` to `adjustIndentation` when default mode is `asIs` when the completion spans across multiple lines.

Reply:
Great! `adjustIndentation` should fix some indentation issues I've seen for postfix before. Looking forward to this improvement!